docs/RFCS/20230731_region_liveness.md

```diff
@@ -0,0 +1,363 @@
+
+The CRDB commit protocol contains a hazard that makes it easy to violate the
+constraint that `txn_timestamp + system.sqlliveness.ttl < unavailable_at`. When
+committing a transaction, if the timestamp cache for a row comes after the
+transaction timestamp, the commit process will refresh the reads and push the
+commit timestamp into the future. This can be avoided by setting a commit
+deadline equal to the transaction timestamp in the transaction that writes the
+`unavailable_at` timestamp. That way if the transaction would push the commit
+timestamp, committing the transaction will fail. The code attempting to set
+`unavailable_at` can retry the transaction, but will need to pick a new
+`unavailabe_at` timestamp.
+
+### Using the KV API
+
+Because `system.sqlliveness`, `system.sqlinstances`, and `systemleases` are
+involved in bootstrapping the SQL layer, the subsystems interacting with the
+tables use the raw KV API instead of the SQL execution layer. Likewise code
+writing to the new `system.region_liveness` table will use the KV API directly.
+It is okay to use SQL to interact with the `system.region_liveness` table as
+long as the SQL is not used during the SQL bootstrap process.
+
+### Why `unavailable_at` Instead of `expiration`?
+
+An alternative design to tracking an `unavailable_at` for each region is the
+table could contain an `expiration` timestamp. SQL servers would periodically
+advance the `expiration` timestamp and it would be used as an upper bound for
+`system.sqlliveness` `expiration` column. The `expiration` design mirrors the
+behavior of the `systemsqlliveness table` and is a little easier to reason about. 
+
+The problem with heart beating an `expiration` column in
+`system.region_liveness` is SQL servers are not always running. So the
+`expiration` column would always be in the past during the cold start process.
+Advancing `expiration` in the cold start process would require a cross-region
+write, which defeats the whole purpose of making the tables regional by row and
+survive zone.
+
+### Why Use a Global Table Instead of Regional by Table?
+
+Most systems interact with the `system.region_liveness` table via reads. Making
+`system.region_liveness` a global table allows for transactions to read from a
+local replica. It would be possible to make `system.region_liveness` regional
+by table if the `system.sqlliveness` heart beating process used stale reads.
+But using stale reads makes the correctness more difficult to reason about.
+
+### When is a Region Marked as Unavailable?
+
+The interface for the `system.region_liveness` table will contain a
+`ProbeLiveness(region string)` function that attempts to read from the
+`system.sqlinstance` table in the probed region. If a region can not be reached
+within 15 seconds, `unavailalbe_at` is set to start the process of fencing out
+the region.
+
+15 seconds is chosen because it gives enough time for leases to transfer in the
```

Comment:
Is there any setting this number is based on?

Reply:
I think the value we care about here is COCKROACH_RANGE_LEASE_DURATION. https://github.com/cockroachdb/cockroach/blob/126274a3ba3f6e41f13676b4fdeb3646350572cb/pkg/base/config.go#L192.

But the sql server's may not have the kv server's configuration for that value.